<a href="https://colab.research.google.com/github/AliAhmed354/Final-Year-Project/blob/main/FYPBreastCancer.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import pandas as pd


In [3]:

dataf = pd.read_csv('/content/drive/MyDrive/Breast_Cancer_Mortality_Survival_Dataset.csv')
dataf.head(5)
dataf.info()
dataf.describe()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4024 entries, 0 to 4023
Data columns (total 21 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Patient ID              4024 non-null   object 
 1   Moth of Birth           4024 non-null   int64  
 2   Age                     4015 non-null   float64
 3   Sex                     3947 non-null   object 
 4   Race                    4024 non-null   object 
 5   Marital Status          4024 non-null   object 
 6   Occupation Code         4021 non-null   float64
 7   Adopted Status          4024 non-null   int64  
 8   T Stage                 4024 non-null   object 
 9   N Stage                 4024 non-null   object 
 10  6th Stage               4024 non-null   object 
 11  differentiate           4024 non-null   object 
 12  Grade                   4024 non-null   int64  
 13  A Stage                 4024 non-null   object 
 14  Tumor Size              4021 non-null   

,Moth of Birth,Age,Occupation Code,Adopted Status,Grade,Tumor Size,Survival Months
count,4024.000000,4015.000000,4021.000000,4024.000000,4024.000000,4021.000000,4024.000000
mean,6.481362,54.094396,8.427257,0.496769,2.150596,30.419299,71.472167
std,3.475442,11.417118,4.637345,0.500052,0.638234,21.161080,25.361855
min,1.000000,30.000000,1.000000,0.000000,1.000000,-75.000000,1.000000
25%,3.000000,47.000000,4.000000,0.000000,2.000000,16.000000,56.000000
50%,6.000000,54.000000,8.000000,0.000000,2.000000,25.000000,73.000000
75%,10.000000,61.000000,12.000000,1.000000,3.000000,38.000000,90.000000
max,12.000000,502.000000,16.000000,1.000000,4.000000,140.000000,760.000000


dataf.info()

In [4]:
dataf.head(5)

,Patient ID,Moth of Birth,Age,Sex,Race,Marital Status,Occupation Code,Adopted Status,T Stage,N Stage,...,differentiate,Grade,A Stage,Tumor Size,Estrogen Status,Progesterone Status,Regional Node Examined,Reginol Node Positive,Survival Months,Mortality_Status
0,A0012,12,68.0,Female,White,Married,11.0,1,T1,N1,...,Poorly differentiated,3,Regional,4.0,Positive,Positive,24,1,60,Alive
1,A0013,12,50.0,Female,White,Married,11.0,0,T2,N2,...,Moderately differentiated,2,Regional,35.0,Positive,Positive,14,5,62,Alive
2,A0014,11,58.0,Female,White,Divorced,9.0,0,T3,N3,...,Moderately differentiated,2,Regional,63.0,Positive,Positive,14,7,75,Alive
3,A0015,3,58.0,Female,White,Married,10.0,0,T1,N1,...,Poorly differentiated,3,Regional,18.0,Positive,Positive,2,1,84,Alive
4,A0016,1,47.0,Female,White,Married,12.0,0,T2,N1,...,Poorly differentiated,3,Regional,41.0,Positive,Positive,3,1,50,Alive


- we have missing values
- some variables are not relevent and can be removed
- variables can be converted to numerical data types



In [5]:
dataf.isnull().sum()

,0
Patient ID,0
Moth of Birth,0
Age,9
Sex,77
Race,0
Marital Status,0
Occupation Code,3
Adopted Status,0
T Stage,0
N Stage,0


In [6]:
dataf.describe()

,Moth of Birth,Age,Occupation Code,Adopted Status,Grade,Tumor Size,Survival Months
count,4024.000000,4015.000000,4021.000000,4024.000000,4024.000000,4021.000000,4024.000000
mean,6.481362,54.094396,8.427257,0.496769,2.150596,30.419299,71.472167
std,3.475442,11.417118,4.637345,0.500052,0.638234,21.161080,25.361855
min,1.000000,30.000000,1.000000,0.000000,1.000000,-75.000000,1.000000
25%,3.000000,47.000000,4.000000,0.000000,2.000000,16.000000,56.000000
50%,6.000000,54.000000,8.000000,0.000000,2.000000,25.000000,73.000000
75%,10.000000,61.000000,12.000000,1.000000,3.000000,38.000000,90.000000
max,12.000000,502.000000,16.000000,1.000000,4.000000,140.000000,760.000000


- max age too high
- -75 can't be a tumor size
- 760 months an outlier




In [7]:
df = dataf.drop(['Patient ID ','Moth of Birth', 'Sex', 'Marital Status', 'Occupation Code', 'Adopted Status'], axis=1, inplace = False )
df.head()


,Age,Race,T Stage,N Stage,6th Stage,differentiate,Grade,A Stage,Tumor Size,Estrogen Status,Progesterone Status,Regional Node Examined,Reginol Node Positive,Survival Months,Mortality_Status
0,68.0,White,T1,N1,IIA,Poorly differentiated,3,Regional,4.0,Positive,Positive,24,1,60,Alive
1,50.0,White,T2,N2,IIIA,Moderately differentiated,2,Regional,35.0,Positive,Positive,14,5,62,Alive
2,58.0,White,T3,N3,IIIC,Moderately differentiated,2,Regional,63.0,Positive,Positive,14,7,75,Alive
3,58.0,White,T1,N1,IIA,Poorly differentiated,3,Regional,18.0,Positive,Positive,2,1,84,Alive
4,47.0,White,T2,N1,IIB,Poorly differentiated,3,Regional,41.0,Positive,Positive,3,1,50,Alive


In [8]:
df = df.dropna()
df.isnull().sum()

,0
Age,0
Race,0
T Stage,0
N Stage,0
6th Stage,0
differentiate,0
Grade,0
A Stage,0
Tumor Size,0
Estrogen Status,0


In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4007 entries, 0 to 4023
Data columns (total 15 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Age                     4007 non-null   float64
 1   Race                    4007 non-null   object 
 2   T Stage                 4007 non-null   object 
 3   N Stage                 4007 non-null   object 
 4   6th Stage               4007 non-null   object 
 5   differentiate           4007 non-null   object 
 6   Grade                   4007 non-null   int64  
 7   A Stage                 4007 non-null   object 
 8   Tumor Size              4007 non-null   float64
 9   Estrogen Status         4007 non-null   object 
 10  Progesterone Status     4007 non-null   object 
 11  Regional Node Examined  4007 non-null   object 
 12  Reginol Node Positive   4007 non-null   object 
 13  Survival Months         4007 non-null   int64  
 14  Mortality_Status        4007 non-null   objec

In [10]:
df['T Stage '].unique()
df['N Stage'].unique()
df['differentiate'].unique()
df['A Stage'].unique()


array(['Regional', 'Distant'], dtype=object)

In [11]:
df['T Stage ']= df['T Stage '].map({'T1':1, 'T2':2, 'T3':3, 'T4':4})


In [12]:
df['N Stage'] = df['N Stage'].map({'N1':1, 'N2':2, 'N3':3})

In [13]:
df['differentiate'] = df['differentiate'].map({'Poorly differentiated':1, 'Moderately differentiated':2,'Well differentiated':3, 'Undifferentiated':4})



In [14]:
df['A Stage'] = df['A Stage'].map({'Regional':1, 'Distant':2})


In [15]:
df['Estrogen Status'] = df['Estrogen Status'].map({'Negative':1, 'Positive':2})

In [16]:
df['Progesterone Status'] = df['Progesterone Status'].map({'Negative':1, 'Positive':2})


In [17]:
df['6th Stage'] = df['6th Stage'].map({'IIA':1, 'IIB':2, 'IIIA':3, 'IIIB':4, 'IIIC':5 })

In [18]:
df['Mortality_Status'] = df['Mortality_Status'].map({'Alive':0, 'Dead':1})

In [19]:
from sklearn import preprocessing
label_encoder = preprocessing.LabelEncoder()

df['Race'] = label_encoder.fit_transform(df['Race'])

In [20]:
df[['Regional Node Examined', 'Reginol Node Positive']] = df[['Regional Node Examined', 'Reginol Node Positive']].apply(pd.to_numeric,errors='coerce')
df.dropna(inplace=True)
df[['Regional Node Examined', 'Reginol Node Positive']] = df[['Regional Node Examined', 'Reginol Node Positive']].apply(pd.to_numeric,errors='coerce')

df.dropna(inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3990 entries, 0 to 4023
Data columns (total 15 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Age                     3990 non-null   float64
 1   Race                    3990 non-null   int64  
 2   T Stage                 3990 non-null   int64  
 3   N Stage                 3990 non-null   int64  
 4   6th Stage               3990 non-null   int64  
 5   differentiate           3990 non-null   int64  
 6   Grade                   3990 non-null   int64  
 7   A Stage                 3990 non-null   int64  
 8   Tumor Size              3990 non-null   float64
 9   Estrogen Status         3990 non-null   int64  
 10  Progesterone Status     3990 non-null   int64  
 11  Regional Node Examined  3990 non-null   float64
 12  Reginol Node Positive   3990 non-null   float64
 13  Survival Months         3990 non-null   int64  
 14  Mortality_Status        3990 non-null   float

In [21]:
df.describe()

,Age,Race,T Stage,N Stage,6th Stage,differentiate,Grade,A Stage,Tumor Size,Estrogen Status,Progesterone Status,Regional Node Examined,Reginol Node Positive,Survival Months,Mortality_Status
count,3990.000000,3990.000000,3990.000000,3990.000000,3990.000000,3990.000000,3990.000000,3990.000000,3990.000000,3990.000000,3990.000000,3990.000000,3990.000000,3990.000000,3990.000000
mean,54.118546,1.776441,1.782456,1.437845,2.319549,1.868421,2.150627,1.022807,30.311028,1.933584,1.827068,14.635840,4.154135,71.473434,0.152130
std,11.424191,0.563489,0.764133,0.693429,1.267008,0.642741,0.638545,0.149307,20.983317,0.249039,0.378236,18.712734,5.111693,25.367723,0.359192
min,30.000000,0.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,-75.000000,1.000000,1.000000,1.000000,1.000000,1.000000,0.000000
25%,47.000000,2.000000,1.000000,1.000000,1.000000,1.000000,2.000000,1.000000,16.000000,2.000000,2.000000,9.000000,1.000000,56.000000,0.000000
50%,54.000000,2.000000,2.000000,1.000000,2.000000,2.000000,2.000000,1.000000,25.000000,2.000000,2.000000,14.000000,2.000000,73.000000,0.000000
75%,61.000000,2.000000,2.000000,2.000000,3.000000,2.000000,3.000000,1.000000,38.000000,2.000000,2.000000,19.000000,5.000000,90.000000,0.000000
max,502.000000,2.000000,4.000000,3.000000,5.000000,4.000000,4.000000,2.000000,140.000000,2.000000,2.000000,1080.000000,46.000000,760.000000,1.000000


In [27]:
import matplotlib.pyplot as plt

cols = ['Age', 'Tumor Size', 'Regional Node Examined', 'Reginol Node Positive', 'Survival Months']

q1 = df[cols].quantile(0.25)
q3 = df[cols].quantile(0.75)
iqr = q3 - q1

lower = q1 - 5 * iqr
upper = q3 + 5 * iqr

outliers = ~((df[cols] < lower) | (df[cols] > upper)).any(axis=1)

newdf = df[outliers]

newdf = newdf[newdf['Tumor Size'] !=-75]

newdf.describe()



,Age,Race,T Stage,N Stage,6th Stage,differentiate,Grade,A Stage,Tumor Size,Estrogen Status,Progesterone Status,Regional Node Examined,Reginol Node Positive,Survival Months,Mortality_Status
count,3944.000000,3944.000000,3944.000000,3944.000000,3944.000000,3944.000000,3944.000000,3944.000000,3944.000000,3944.000000,3944.000000,3944.000000,3944.000000,3944.000000,3944.000000
mean,53.997211,1.775609,1.778905,1.420385,2.289807,1.869168,2.150101,1.021552,30.234280,1.934331,1.828093,14.165822,3.874239,71.443205,0.148834
std,8.964423,0.564486,0.761621,0.677339,1.241941,0.643446,0.639224,0.145233,20.823556,0.247735,0.377347,7.883850,4.360546,22.819925,0.355970
min,30.000000,0.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,0.000000
25%,47.000000,2.000000,1.000000,1.000000,1.000000,1.000000,2.000000,1.000000,16.000000,2.000000,2.000000,9.000000,1.000000,56.000000,0.000000
50%,54.000000,2.000000,2.000000,1.000000,2.000000,2.000000,2.000000,1.000000,25.000000,2.000000,2.000000,14.000000,2.000000,73.000000,0.000000
75%,61.000000,2.000000,2.000000,2.000000,3.000000,2.000000,3.000000,1.000000,38.000000,2.000000,2.000000,19.000000,5.000000,90.000000,0.000000
max,69.000000,2.000000,4.000000,3.000000,5.000000,4.000000,4.000000,2.000000,140.000000,2.000000,2.000000,61.000000,25.000000,107.000000,1.000000


In [49]:
from sklearn.pipeline import Pipeline
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import f1_score, accuracy_score, confusion_matrix
## defining x and y vales, train test split

xvalues = newdf.drop(columns=['Mortality_Status'])
yvalues = newdf['Mortality_Status']

xtrain, xtest, ytrain, ytest = train_test_split(
    xvalues, yvalues, test_size=0.2, random_state=42, stratify=yvalues
)

## set models with parameters

models = {
    'rforrest': Pipeline([
        ('scaler', StandardScaler()),
        ('model', RandomForestClassifier(n_estimators=100, random_state=42))
    ]),

    'svm': Pipeline([
        ('scaler', StandardScaler()),
        ('model', SVC(kernel='rbf', random_state=42))
    ]),

    'knn': Pipeline([
        ('scaler', StandardScaler()),
        ('model', KNeighborsClassifier(n_neighbors=10))
    ])
}

## set resampling techniques

from imblearn.over_sampling import SMOTE
from imblearn.under_sampling import RandomUnderSampler
from imblearn.combine import SMOTETomek

resamplers = {
    'none': None,
    'smote':SMOTE(random_state=42),
    'unders':RandomUnderSampler(random_state=42),
    'smotetomek':SMOTETomek(random_state=42)
}




In [50]:
## looping through each combo

results = []
confmatrix = []

count = 0

for rname, resampler in resamplers.items():
  for mname, pipeline in models.items():

    xt_result, yt_result = xtrain, ytrain

    if resampler is not None:
      xt_result, yt_result = resampler.fit_resample(xtrain, ytrain)

    pipeline.fit(xt_result, yt_result)
    ypred = pipeline.predict(xtest)

    f1 = f1_score(ytest, ypred)
    acc = accuracy_score(ytest, ypred)

    tn, fp, fn, tp = confusion_matrix(ytest, ypred).ravel()

    results.append({
        'count': count,
        'resampler':rname,
        'model':mname,
        'f1':f1,
        'accuracy':acc})

    confmatrix.append({
        'count' : count,
        'resampler':rname,
        'model':mname,
        'tn':tn,
        'fp':fp,
        'fn':fn,
        'tp':tp
    })

    count += 1







In [52]:
resultsdf = pd.DataFrame(results)
resultsdf.head(20)

,count,resampler,model,f1,accuracy
0,0,none,rforrest,0.612903,0.908745
1,1,none,svm,0.497041,0.892269
2,2,none,knn,0.373333,0.880862
3,3,smote,rforrest,0.622010,0.899873
4,4,smote,svm,0.595556,0.884664
5,5,smote,knn,0.454545,0.847909
6,6,unders,rforrest,0.516717,0.798479
7,7,unders,svm,0.534884,0.797212
8,8,unders,knn,0.517483,0.825095
9,9,smotetomek,rforrest,0.613208,0.896071


In [53]:
## lin regression for survival months

regx = newdf.drop(columns=['Survival Months'])
regy = newdf['Survival Months']


from sklearn.linear_model import LinearRegression

linxtrain, linxtest, linytrain, linytest = train_test_split(
    regx, regy, test_size=0.2, random_state=42
)

linreg = LinearRegression()
linreg.fit(linxtrain, linytrain)

linpred = linreg.predict(linxtest)


In [54]:
## random forrest for regression

from sklearn.ensemble import RandomForestRegressor

rfxtrain, rfxtest, rfytrain, rftest = train_test_split(
    regx, regy, test_size=0.2, random_state=42
)

rfr = RandomForestRegressor(n_estimators=100, random_state=42)
rfr.fit(rfxtrain, rfytrain)

rfrpred = rfr.predict(rfxtest)

In [57]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

linmae = mean_absolute_error(linytest, linpred)
linmse = mean_squared_error(linytest, linpred)
linr2 = r2_score(linytest, linpred)

rfrmae = mean_absolute_error(rftest, rfrpred)
rfrmse = mean_squared_error(rftest, rfrpred)
rfr2 = r2_score(rftest, rfrpred)

print(rfr2)

0.21040765840049336
